src/main_frontend.cpp

```diff
@@ -0,0 +1,614 @@
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <stdio.h>
+#include <string.h>
+#define SDL_MAIN_HANDLED
+#include "SDL.h"
+#include "mcu.h"
+#include "mcu_opcodes.h"
+#include "mcu_interrupt.h"
+#include "mcu_timer.h"
+#include "pcm.h"
+#include "lcd.h"
+#include "submcu.h"
+#include "midi.h"
+#include "utf8main.h"
+#include "utils/files.h"
+#include "emu.h"
+#include "ringbuffer.h"
+#include "math_util.h"
+
+#if __linux__
+#include <unistd.h>
+#include <limits.h>
+#endif
+
+struct fe_emu_instance_t {
+    emu_t        emu;
+    ringbuffer_t sample_buffer;
+    SDL_Thread*  thread;
+    bool         running;
+};
+
+const size_t FE_MAX_INSTANCES = 16;
+
+struct frontend_t {
+    fe_emu_instance_t instances[FE_MAX_INSTANCES];
+    size_t instances_in_use = 0;
+
+    int audio_buffer_size = 0;
+    int audio_page_size = 0;
+
+    SDL_AudioDeviceID sdl_audio = 0;
+};
+
+bool FE_AllocateInstance(frontend_t& container, fe_emu_instance_t** result)
+{
+    if (container.instances_in_use == FE_MAX_INSTANCES)
+    {
+        return false;
+    }
+
+    fe_emu_instance_t& fe = container.instances[container.instances_in_use];
+    memset(&fe, 0, sizeof(fe_emu_instance_t));
+    ++container.instances_in_use;
+
+    if (result)
+    {
+        *result = &fe;
+    }
+
+    return true;
+}
+
+void FE_SendMIDI(frontend_t& fe, size_t n, uint8_t* first, uint8_t* last)
+{
+    while (first != last)
+    {
+        MCU_PostUART(*fe.instances[n].emu.mcu, *first);
+        ++first;
+    }
+}
+
+void FE_BroadcastMIDI(frontend_t& fe, uint8_t* first, uint8_t* last)
+{
+    for (size_t i = 0; i < fe.instances_in_use; ++i)
+    {
+        FE_SendMIDI(fe, i, first, last);
+    }
+}
+
+void FE_RouteMIDI(frontend_t& fe, uint8_t* first, uint8_t* last)
+{
+    if (*first < 0x80)
+    {
+        printf("FE_RouteMIDI received data byte %02x\n", *first);
+        return;
+    }
+
+    const bool is_sysex = *first == 0xF0;
+    const uint8_t channel = *first & 0x0F;
+
+    if (is_sysex)
+    {
+        FE_BroadcastMIDI(fe, first, last);
+    }
+    else
+    {
+        FE_SendMIDI(fe, channel % fe.instances_in_use, first, last);
+    }
+}
+
+void FE_ReceiveSample(void* userdata, int *sample)
```

Comment:
Passing something like an `AudioFrame` struct we use in DOSBox Staging would be nicer than these raw unsafe pointers:

https://github.com/dosbox-staging/dosbox-staging/blob/main/include/audio_frame.h

Another general comment: using `std::array` or `std::vector` is zero-cost compared to C arrays but safer and there's potential to handle them in a nicer way (e.g., using iterators instead of error prone indexing and oldschool for loops).

Another

Reply:
Yes, I think this function should be changed (as well as MCU_PostSample) but I left it as I found it. The `int*` is a performance pessimization and it's not an intuitive API.

> using std::array or std::vector is zero-cost compared to C arrays but safer

I totally agree with you and I would have written it that way myself, but I didn't want to introduce a bunch of C++ features since the current code is not using them. As far as I can tell the only STL type used anywhere is std::string, and std::vector made it into the rtmidi module only because it's required for the rtmidi callback.

Reply:
> performance pessimization and it's not an intuitive API

Haha, okay, I can see I'm among friends—we're 100% on the same page 😏 I get it, you didn't want to change too much.